django/db/models/sql/datastructures.py

```diff
@@ -160,9 +171,17 @@ def as_sql(self, compiler, connection):
     def relabeled_clone(self, change_map):
         return self.__class__(self.table_name, change_map.get(self.table_alias, self.table_alias))
 
+    @property
+    def identity(self):
+        return self.__class__, self.table_name, self.table_alias
```

Comment:
Same here...

```suggestion
        return self.table_name, self.table_alias
```

Reply:
Adding `self.__class__` is important in the `__hash__` case other wise a subclass of `BaseTable` or `Join` would hash to the same value.